text/0000-return-position-impl-trait-in-traits.md

```diff
@@ -0,0 +1,322 @@
+
+The [future possibilities](#future-possibilities) section discusses a planned extension to support naming the type returned by an impl trait, which could work to overcome this limitation for clients.
+
+## Impls cannot add new methods to the resulting types
+
+Similarly to the previous point, [@Gankra also pointed out][gankra] that impls which employ `-> impl Trait` are not able to add methods or implement traits on the resulting types. With `IntoIterator`, one has the option of adding inherent (or trait) methods to the resulting type:
+
+```rust
+struct MyIterator {
+    ...
+}
+
+impl IntoIterator for MyType {
+    type IntoIter = MyIterator;
+    ...
+}
+
+impl MyIterator {
+    fn extra_method(&self) {
+
+    }
+}
+```
+
+Using `-> impl Trait` tends to make that not work. One way to address this would be to permit impls to specify the return types of `-> impl Trait` as concrete types.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+## Can traits migrate from a named associated type to `impl Trait`?
+
+Not compatibly, no, because they would no longer have a named associated type.
+
+## Can traits migrate from `impl Trait` to a named associated type?
+
+Generally yes, but all impls would have to be rewritten.
+
+## Would there be any way to make it possible to migrate from `impl Trait` to a named associated type compatibly?
+
+Potentially! There have been proposals to allow the values of associated types that appear in function return types to be inferred from the function declaration. So the trait has `fn method(&self) -> Self::Iter` and the impl has `fn method(&self) -> impl Iterator`, then the impl would also be inferred to have `type Iter = impl Iterator` (and the return type rewritten to reference it). This may be a good idea, but it is not proposed as part of this RFC.
+
+## What about using a named associated type?
+
+One alternative under consideration was to use a named associated type instead of the anonymous `$` type. The name could be derived by converting "snake case" methods to "camel case", for example. This has the advantage that users of the trait can refer to the return type by name.
+
+We decided against this proposal:
+
+* Introducing a name by converting to camel-case feels surprising and inelegant.
+* Return position impl Trait in other kinds of functions doesn't introduce any sort of name for the return type, so it is not analogous.
+
+There is a need to introduce a mechanism for naming the return type for functions that use `-> impl Trait`; we plan to introduce a second RFC addressing this need uniformly across all kinds of functions.
```

Comment:
Side note: unsure what all is being discussed in the lead up to this RFC but something I've been playing with in my mind is functions also being types.  This would allow `Trait::func::ReturnType`.

No idea what has been discussed for a desugared version of `-> dyn Trait` but the above idea came up as part my "how would I implement it" brainstorming.  a `-> dyn Trait `would desugar to `Trait::func::alloc_requirements` (size, alignment) and `Trait::func::func_into` (wrapper around the user's function, moving into the provided memory)